pkg/core/transaction/transaction.go

```diff
@@ -233,3 +238,194 @@ func (t *Transaction) Bytes() []byte {
+	case InvocationType:
+		tx.Script = hex.EncodeToString(t.Data.(*InvocationTX).Script)
+		tx.Gas = t.Data.(*InvocationTX).Gas
+	case PublishType:
+		transaction := t.Data.(*PublishTX)
+		tx.Contract = &contract{
+			Code: code{
+				Hash:       hash.Hash160(transaction.Script),
+				Script:     hex.EncodeToString(transaction.Script),
+				ParamList:  transaction.ParamList,
+				ReturnType: transaction.ReturnType,
+			},
+			NeedStorage: transaction.NeedStorage,
+			Name:        transaction.Name,
+			CodeVersion: transaction.CodeVersion,
+			Author:      transaction.Author,
+			Email:       transaction.Email,
+			Description: transaction.Description,
+		}
+	case RegisterType:
+		transaction := *t.Data.(*RegisterTX)
+		tx.Asset = &asset{
+			AssetType: transaction.AssetType,
+			Name:      json.RawMessage(transaction.Name),
+			Amount:    transaction.Amount,
+			Precision: transaction.Precision,
+			Owner:     transaction.Owner,
+			Admin:     address.Uint160ToString(transaction.Admin),
+		}
+	case StateType:
+		tx.Descriptors = t.Data.(*StateTX).Descriptors
+	}
+	return json.Marshal(tx)
+}
+
+// UnmarshalJSON implements json.Unmarshaler interface.
+func (t *Transaction) UnmarshalJSON(data []byte) error {
+	tx := new(transactionJSON)
+	if err := json.Unmarshal(data, tx); err != nil {
+		return err
+	}
+	t.Type = tx.Type
+	t.Version = tx.Version
+	t.Attributes = tx.Attributes
+	t.Inputs = tx.Inputs
+	t.Outputs = tx.Outputs
+	t.Scripts = tx.Scripts
+	switch tx.Type {
+	case MinerType:
+		t.Data = &MinerTX{
+			Nonce: tx.Nonce,
+		}
+	case ClaimType:
+		t.Data = &ClaimTX{
+			Claims: tx.Claims,
+		}
+	case EnrollmentType:
+		t.Data = &EnrollmentTX{
+			PublicKey: *tx.PublicKey,
+		}
+	case InvocationType:
+		bytes, err := hex.DecodeString(tx.Script)
+		if err != nil {
+			return err
+		}
+		t.Data = &InvocationTX{
+			Script:  bytes,
+			Gas:     tx.Gas,
+			Version: 1,
+		}
+	case PublishType:
+		bytes, err := hex.DecodeString(tx.Contract.Code.Script)
+		if err != nil {
+			return err
+		}
+		t.Data = &PublishTX{
+			Script:      bytes,
+			ParamList:   tx.Contract.Code.ParamList,
+			ReturnType:  tx.Contract.Code.ReturnType,
+			NeedStorage: tx.Contract.NeedStorage,
+			Name:        tx.Contract.Name,
+			CodeVersion: tx.Contract.CodeVersion,
+			Author:      tx.Contract.Author,
+			Email:       tx.Contract.Email,
+			Description: tx.Contract.Description,
+			Version:     uint8(1),
+		}
+	case RegisterType:
+		admin, err := address.StringToUint160(tx.Asset.Admin)
+		if err != nil {
+			return err
+		}
+		t.Data = &RegisterTX{
+			AssetType: tx.Asset.AssetType,
+			Name:      string(tx.Asset.Name),
+			Amount:    tx.Asset.Amount,
+			Precision: tx.Asset.Precision,
+			Owner:     tx.Asset.Owner,
+			Admin:     admin,
+		}
+	case StateType:
+		t.Data = &StateTX{
+			Descriptors: tx.Descriptors,
+		}
+	case ContractType:
+		t.Data = &ContractTX{}
+	case IssueType:
+		t.Data = &IssueTX{}
+	}
+
+	err := t.createHash()
```

Comment:
`t.Hash()` will do that automatically, so this code is not very useful.

Reply:
Unless of course you're doing some kind of struct reuse and expecting `hash` to be already set.